backends/ubpf/ubpfParser.cpp

```diff
@@ -0,0 +1,361 @@
+            builder->emitIndent();
+            visit(expr);
+            builder->appendFormat(".%s = (", field.c_str());
+            type->emit(builder);
+            builder->appendFormat(")((%s(%s, BYTES(%s))",
+                                  helper,
+                                  program->packetStartVar.c_str(),
+                                  program->offsetVar.c_str());
+            if (shift != 0)
+                builder->appendFormat(" >> %d", shift);
+            builder->append(")");
+
+            if (widthToExtract != loadSize) {
+                builder->append(" & BPF_MASK(");
+                type->emit(builder);
+                builder->appendFormat(", %d)", widthToExtract);
+            }
+
+            builder->append(")");
+            builder->endOfStatement(true);
+        } else {
+            // wide values; read all bytes one by one.
+            unsigned shift;
+            if (alignment == 0)
+                shift = 0;
+            else
+                shift = 8 - alignment;
+
+            const char* helper;
+            if (shift == 0)
+                helper = "load_byte";
+            else
+                helper = "load_half";
+            auto bt = UBPFTypeFactory::instance->create(IR::Type_Bits::get(8));
+            unsigned bytes = ROUNDUP(widthToExtract, 8);
+            for (unsigned i=0; i < bytes; i++) {
+                builder->emitIndent();
+                visit(expr);
+                builder->appendFormat(".%s[%d] = (", field.c_str(), i);
+                bt->emit(builder);
+                builder->appendFormat(")((%s(%s, BYTES(%s) + %d) >> %d)",
+                                      helper,
+                                      program->packetStartVar.c_str(),
+                                      program->offsetVar.c_str(), bytes - i - 1, shift);
+
+                if ((i == bytes - 1) && (widthToExtract % 8 != 0)) {
+                    builder->append(" & BPF_MASK(");
+                    bt->emit(builder);
+                    builder->appendFormat(", %d)", widthToExtract % 8);
+                }
+
+                builder->append(")");
+                builder->endOfStatement(true);
+            }
+        }
+
+        builder->emitIndent();
+        builder->appendFormat("%s += %d", program->offsetVar.c_str(), widthToExtract);
+        builder->endOfStatement(true);
+        builder->newline();
+    }
+
+    void
+    UBPFStateTranslationVisitor::compileExtract(const IR::Expression* destination) {
+        auto type = state->parser->typeMap->getType(destination);
+        auto ht = type->to<IR::Type_StructLike>();
+
+        if (ht == nullptr) {
+            ::error("Cannot extract to a non-struct type %1%", destination);
+            return;
+        }
+
+        unsigned width = ht->width_bits();
+        auto program = state->parser->program;
+        builder->emitIndent();
+        // TODO: Check if packet is not too short
```

Comment:
This should not be too difficult... can you please do this?

Reply:
That TODO is not outdated (to be removed). 

The packet length's check is implemented in [1].

[1] https://github.com/p4lang/p4c/blob/9b6868a0d0eb48d3b2dced4e43ada1aa20636e9b/backends/ubpf/ubpfProgram.cpp#L278